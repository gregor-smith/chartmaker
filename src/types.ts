import {
    Record as Record_,
    Literal,
    String,
    Union,
    InstanceOf,
    Boolean,
    Static,
    Number as Number_,
    Array as Array_,
    Runtype,
    Partial as Partial_
} from 'runtypes'

import {
    CHART_ALBUMS_COUNT,
    MAX_SCREENSHOT_SCALE,
    MAX_COLLAGE_ROWS_X,
    MAX_COLLAGE_ROWS_Y,
    STATE_VERSION,
    EXPORT_CHART_PLACEHOLDER
} from '@/constants'


const Integer = Number_.withConstraint(Number.isSafeInteger)


function IntegerRange(minimum: number, maximum = Number.MAX_SAFE_INTEGER) {
    return Integer.withConstraint(number => number >= minimum && number <= maximum)
}


function FixedSizeArray<T extends Runtype>(element: T, size: number) {
    return Array_(element).withConstraint(array => array.length === size)
}


function NonEmptyArray<T extends Runtype>(element: T) {
    return Array_(element).withConstraint(array => array.length > 0)
}


const NonEmptyString = String.withConstraint(string => string.length !== 0)


const PositiveInteger = IntegerRange(0)


const V1PlaceholderAlbum = Record_({
    placeholder: Literal(true),
    id: PositiveInteger
})


const SearchAlbum = Record_({
    name: NonEmptyString,
    url: NonEmptyString
})


const V1NamedAlbum = SearchAlbum.And(
    Record_({
        placeholder: Literal(false),
        id: PositiveInteger
    })
)
const V2NamedAlbum = SearchAlbum.And(
    Record_({
        id: PositiveInteger
    })
)
export const NamedAlbum = V2NamedAlbum


const V1Album = V1NamedAlbum.Or(V1PlaceholderAlbum)
const V2Album = V2NamedAlbum.Or(PositiveInteger)
const Album = V2Album


const ChartShape = Union(
    Record_({
        tag: Literal('Top'),
        size: Union(
            Literal(40),
            Literal(42),
            Literal(100)
        )
    }),
    Record_({
        tag: Literal('Collage')
    })
)


const V1Chart = Record_({
    name: NonEmptyString,
    albums: FixedSizeArray(V1Album, CHART_ALBUMS_COUNT),
    shape: ChartShape,
    rowsX: IntegerRange(1, MAX_COLLAGE_ROWS_X),
    rowsY: IntegerRange(1, MAX_COLLAGE_ROWS_Y)
})
const V2Chart = Record_({
    ...V1Chart.fields,
    albums: FixedSizeArray(Album, CHART_ALBUMS_COUNT)
})
const Chart = V2Chart


const SearchState = Union(
    Record_({
        tag: Literal('Waiting')
    }),
    Record_({
        tag: Literal('Loading'),
        controller: InstanceOf(AbortController)
    }),
    Record_({
        tag: Literal('Complete'),
        albums: NonEmptyArray(SearchAlbum)
    }),
    Record_({
        tag: Literal('Error'),
        message: String
    })
).And(
    Record_({
        query: String
    })
)


const ScreenshotState = Record_({
    loading: Boolean,
    scale: IntegerRange(1, MAX_SCREENSHOT_SCALE)
})


const ExportChartAlbum = SearchAlbum.Or(Literal(EXPORT_CHART_PLACEHOLDER))

const ExportChartShape = ChartShape.alternatives[0].Or(
    ChartShape.alternatives[1].And(
        Record_({
            rowsX: V2Chart.fields.rowsX,
            rowsY: V2Chart.fields.rowsY,
        })
    )
)

const ExportChart = Record_({
    name: V2Chart.fields.name,
    albums: FixedSizeArray(ExportChartAlbum, CHART_ALBUMS_COUNT),
    shape: ExportChartShape
})


export const V1State = Record_({
    apiKey: String,
    charts: NonEmptyArray(V1Chart),
    activeChartIndex: PositiveInteger,
    search: SearchState,
    screenshot: ScreenshotState
}).And(
    Partial_({
        version: Literal(1),
        highlightedID: PositiveInteger
    })
)
export const V2State = Record_({
    ...V1State.intersectees[0].fields,
    version: Literal(2),
    charts: NonEmptyArray(V2Chart)
}).And(
    Partial_({
        highlightedID: PositiveInteger
    })
)
const V3State = Record_({
    ...V2State.intersectees[0].fields,
    version: Literal(STATE_VERSION)
}).And(
    Partial_({
        ...V2State.intersectees[1].fields,
        viewing: ExportChart
    })
)
export const State = V3State


export type NamedAlbum = Static<typeof NamedAlbum>
export type SearchAlbum = Static<typeof SearchAlbum>
export type Album = Static<typeof Album>
export type ChartShape = Static<typeof ChartShape>
export type Chart = Static<typeof Chart>
export type SearchState = Static<typeof SearchState>
export type ScreenshotState = Static<typeof ScreenshotState>
export type State = Static<typeof State>
export type V1State = Static<typeof V1State>
export type V2State = Static<typeof V2State>
export type ExportChart = Static<typeof ExportChart>
export type ExportChartShape = Static<typeof ExportChartShape>
export type ExportChartAlbum = Static<typeof ExportChartAlbum>
