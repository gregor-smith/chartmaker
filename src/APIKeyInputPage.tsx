import React, { FC, ChangeEvent, FormEvent } from 'react'
import { css } from 'emotion'

import { DispatchProps } from './reducer'


export type APIKeyInputPageState = {
    key: string
}


type Props = DispatchProps<'UpdateAPIKeyInputText' | 'ChangeAPIKey'> & {
    text: string
}


const pageStyle = css({
    display: 'flex',
    width: '100%',
    height: '100%',
    alignItems: 'center',
    justifyContent: 'center',
    flexGrow: 1
})


const formStyle = css({
    display: 'flex',
    padding: '1rem',
    border: '1px solid white'
})


export const APIKeyInputPage: FC<Props> = ({ dispatch, text: key }) => {
    function updateKey(event: ChangeEvent<HTMLInputElement>) {
        event.preventDefault()
        dispatch({
            tag: 'UpdateAPIKeyInputText',
            key: event.currentTarget.value
        })
    }

    function confirmKey(event: FormEvent<HTMLFormElement>) {
        event.preventDefault()
        dispatch({ tag: 'ChangeAPIKey', key })
    }

    return (
        <main className={pageStyle}>
            <form className={formStyle} onSubmit={confirmKey}>
                <input type='password'
                    placeholder='Enter Last.fm API key'
                    autoFocus
                    onChange={updateKey}/>
                <button type='submit'>
                    Confirm
                </button>
            </form>
        </main>
    )
}
