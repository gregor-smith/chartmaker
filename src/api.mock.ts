import type { searchLastFM as _search } from './api.js'
export type { SearchResult, SearchArguments } from './api.js'


export const search = jest.fn<ReturnType<typeof _search>, Parameters<typeof _search>>()
