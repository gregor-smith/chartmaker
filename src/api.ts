import {
    Record as Record_,
    Array as Array_,
    String as String_,
    Static
} from 'runtypes'


const LastFMAlbum = Record_({
    name: String_,
    artist: String_,
    image: Array_(
        Record_({
            '#text': String_
        })
    )
})


const LastFMResult = Record_({
    results: Record_({
        albummatches: Record_({
            album: Array_(LastFMAlbum)
        })
    })
})


export type LastFMAlbum = Static<typeof LastFMAlbum>
export type LastFMResult = Static<typeof LastFMResult>


function isLastFMNullString(value: string): boolean {
    // Although Last.fm offers a JSON API, it doesn't make use of any JSON
    // types other than strings. Null values are represented by the string
    // '(null)' or just an empty string. Why even bother with JSON then...?
    return value === '' || value === '(null)'
}


export type SearchResultAlbum = {
    name: string
    url: string
}


function formatLastFMResult(result: LastFMResult): SearchResultAlbum[] {
    const albums: SearchResultAlbum[] = []
    for (const album of result.results.albummatches.album) {
        if (isLastFMNullString(album.artist)
                || isLastFMNullString(album.name)
                || album.image.some(image => isLastFMNullString(image['#text']))) {
            continue
        }
        albums.push({
            name: `${album.artist} - ${album.name}`,
            url: album.image[album.image.length - 1]['#text']
        })
    }
    return albums
}


export type SearchArguments = {
    key: string
    query: string
    signal: AbortSignal
}


export type SearchResult =
    | { tag: 'Ok', albums: SearchResultAlbum[] }
    | { tag: 'StatusError', status: number }
    | { tag: 'JSONDecodeError' }
    | { tag: 'InvalidResponseData' }
    | { tag: 'NetworkError' }
    | { tag: 'Cancelled' }


function joinURLQuery(base: string, query: Record<string, string>): string {
    const joinedQuery = Object.entries(query)
        .map(([ key, value ]) => {
            const escapedKey = encodeURIComponent(key.trim())
            const escapedValue = encodeURIComponent(value.trim())
            return `${escapedKey}=${escapedValue}`
        })
        .join('&')
    return `${base}?${joinedQuery}`
}


export async function search({
    key,
    query,
    signal
}: SearchArguments): Promise<SearchResult> {
    const url = joinURLQuery('https://ws.audioscrobbler.com/2.0/', {
        method: 'album.search',
        format: 'json',
        api_key: key,
        album: query
    })

    let response: Response
    try {
        response = await fetch(url, { signal })
    }
    catch {
        return {
            tag: signal.aborted
                ? 'Cancelled'
                : 'NetworkError'
        }
    }
    if (!response.ok) {
        return {
            tag: 'StatusError',
            status: response.status
        }
    }

    let result: unknown
    try {
        result = await response.json()
    }
    catch {
        return { tag: 'JSONDecodeError' }
    }

    if (!LastFMResult.guard(result)) {
        return { tag: 'InvalidResponseData' }
    }

    return {
        tag: 'Ok',
        albums: formatLastFMResult(result)
    }
}
