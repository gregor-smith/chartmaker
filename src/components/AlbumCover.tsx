import React, { FC, DragEvent } from 'react'
import { css, cx } from 'emotion'

import { TEXT_COLOUR, ALBUM_PADDING_SIZE, highlightBackgroundStyle } from '@/style'
import { Album } from '@/types'


export type AlbumCoverProps = {
    album: Album
    size: string
    onDragStart: (event: DragEvent<HTMLDivElement>) => void
    onDragOver?: (event: DragEvent<HTMLDivElement>) => void
    onDragEnter?: (event: DragEvent<HTMLDivElement>) => void
    onDrop?: (event: DragEvent<HTMLDivElement>) => void
    onMouseEnter?: () => void
    overlayClass?: string
    highlighted?: boolean
}


const baseContainerStyle = css({
    position: 'relative',
    background: TEXT_COLOUR,
    margin: ALBUM_PADDING_SIZE
})

const imageStyle = css({
    position: 'absolute',
    top: 0,
    left: 0,
    width: '100%',
    maxWidth: '100%',
    height: '100%',
    objectFit: 'contain'
})

// the overlay is given an additional non-emotion class so it can be used as a
// selector in the container's style
const overlaySelector = 'overlay'

const baseOverlayStyle = cx(
    overlaySelector,
    imageStyle,
    css({ zIndex: 10 })
)


export const AlbumCover: FC<AlbumCoverProps> = ({
    album,
    size,
    onDragStart,
    onDragEnter,
    onDragOver,
    onDrop,
    onMouseEnter,
    children,
    overlayClass,
    highlighted
}) => {
    const overlayStyle = cx(baseOverlayStyle, overlayClass)
    const containerStyle = cx(
        baseContainerStyle,
        css({
            width: size,
            height: size,
            [`:not(:hover) .${overlaySelector}`]: {
                display: 'none'
            }
        }),
        highlighted === false
            ? highlightBackgroundStyle
            : undefined
    )

    let image: JSX.Element | undefined
    if (!album.placeholder) {
        image = (
            // lazy loading causes problems with html2canvas so keep it default
            <img className={imageStyle}
                src={album.url}
                alt={album.name}
                crossOrigin='anonymous'/>
        )
    }

    return (
        <div className={containerStyle}
                draggable={!album.placeholder}
                onDragStart={onDragStart}
                onDragEnter={onDragEnter}
                onDragOver={onDragOver}
                onDrop={onDrop}
                onMouseEnter={onMouseEnter}
                title={album.placeholder ? undefined : album.name}>
            {image}
            <div className={overlayStyle}>
                {children}
            </div>
        </div>
    )
}
