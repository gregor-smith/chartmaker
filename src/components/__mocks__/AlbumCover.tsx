import type { FC } from 'react'
import { cx } from 'emotion'

import type { AlbumCoverProps } from '@/components/AlbumCover'


export const AlbumCover: FC<AlbumCoverProps> = ({
    album,
    size,
    overlayClass,
    children,
    highlighted,
    ...props
}) => {
    const json = JSON.stringify(album)
    const className = cx('mock-album-cover', overlayClass)
    return (
        <div {...props} className={className}>
            {`Album: ${json}`}
            {`Size: ${size}`}
            {`Highlighted: ${highlighted}`}
            {children}
        </div>
    )
}
