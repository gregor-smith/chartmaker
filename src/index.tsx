import 'bootstrap/scss/bootstrap-reboot.scss'
import { render } from 'react-dom'

import { App } from '@/App'


render(
    <App/>,
    document.getElementById('react-root')
)
