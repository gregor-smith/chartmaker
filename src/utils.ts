import html2canvas from 'html2canvas'

import { BACKGROUND_COLOUR } from './style'


export function readFileText(file: File): Promise<string> {
    return new Promise((resolve, reject) => {
        const reader = new FileReader()
        reader.addEventListener('error', reject)
        reader.addEventListener('load', () => resolve(reader.result as string))
        reader.readAsText(file, 'utf-8')
    })
}


export function findIndex<T>(array: ReadonlyArray<T>, predicate: (item: T) => boolean): number | null {
    const index = array.findIndex(predicate)
    return index === -1 ? null : index
}


export async function elementToDataURI(element: HTMLElement, scale: number) {
    const canvas = await html2canvas(element, {
        allowTaint: true,
        scale,
        useCORS: true,
        imageTimeout: 0,
        backgroundColor: BACKGROUND_COLOUR
    })
    return canvas.toDataURL()
}


export function jsonToDataURI(json: string): string {
    return 'data:application/json;charset=utf-8,' + encodeURIComponent(json)
}


export function downloadURI(uri: string, filename: string) {
    const link = document.createElement('a')
    link.href = uri
    link.download = filename
    link.click()
    link.remove()
}
