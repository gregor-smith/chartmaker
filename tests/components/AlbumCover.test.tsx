import React, { createRef, MutableRefObject, DragEvent } from 'react'
import { render } from 'react-dom'
import { css } from 'emotion'
import { act } from 'react-dom/test-utils'

import AlbumCover from '../../src/components/AlbumCover'
import { setRenderContainer, clearRenderContainer, ignore, dragEvent } from '../utils'
import { Album, Size } from '../../src/types'


const container: MutableRefObject<HTMLElement | null> = createRef()
beforeEach(() => setRenderContainer(container))
afterEach(() => clearRenderContainer(container))

const dragMock = jest.fn<void, DragEvent<HTMLDivElement>[]>()
afterEach(jest.clearAllMocks)


const namedAlbum: Album = {
    placeholder: false,
    id: 123,
    name: 'Test album',
    url: 'https://test.com'
}
const placeholderAlbum: Album = {
    placeholder: true,
    id: 123
}


test('renders without image for placeholder album', () => {
    render(
        <AlbumCover album={placeholderAlbum}
                size={'2rem' as Size}
                onDragStart={ignore}>
            Test children
        </AlbumCover>,
        container.current
    )
    expect(container.current).toMatchSnapshot()
})


test('renders with image for named album', () => {
    render(
        <AlbumCover album={namedAlbum}
                size={'2rem' as Size}
                onDragStart={ignore}>
            Test children
        </AlbumCover>,
        container.current
    )
    expect(container.current).toMatchSnapshot()
})


test('overlayClass prop merges styles', () => {
    render(
        <AlbumCover album={placeholderAlbum}
                size={'2rem' as Size}
                onDragStart={ignore}
                overlayClass={css({ color: 'red' })}>
            Test children
        </AlbumCover>,
        container.current
    )
    expect(container.current).toMatchSnapshot()
})


test('dragstart event calls onDragStart prop', () => {
    render(
        <AlbumCover album={namedAlbum}
                size={'2rem' as Size}
                onDragStart={dragMock}>
            Test children
        </AlbumCover>,
        container.current
    )
    act(() => {
        container.current
            ?.firstChild
            ?.dispatchEvent(dragEvent('dragstart'))
    })
    expect(dragMock).toHaveBeenCalledTimes(1)
})


test('dragenter event calls onDragEnter prop', () => {
    render(
        <AlbumCover album={namedAlbum}
                size={'2rem' as Size}
                onDragStart={ignore}
                onDragEnter={dragMock}>
            Test children
        </AlbumCover>,
        container.current
    )
    act(() => {
        container.current
            ?.firstChild
            ?.dispatchEvent(dragEvent('dragenter'))
    })
    expect(dragMock).toHaveBeenCalledTimes(1)
})


test('dragover event calls onDragOver prop', () => {
    render(
        <AlbumCover album={namedAlbum}
                size={'2rem' as Size}
                onDragStart={ignore}
                onDragOver={dragMock}>
            Test children
        </AlbumCover>,
        container.current
    )
    act(() => {
        container.current
            ?.firstChild
            ?.dispatchEvent(dragEvent('dragover'))
    })
    expect(dragMock).toHaveBeenCalledTimes(1)
})


test('drop event calls onDrop prop', () => {
    render(
        <AlbumCover album={namedAlbum}
                size={'2rem' as Size}
                onDragStart={ignore}
                onDrop={dragMock}>
            Test children
        </AlbumCover>,
        container.current
    )
    act(() => {
        container.current
            ?.firstChild
            ?.dispatchEvent(dragEvent('drop'))
    })
    expect(dragMock).toHaveBeenCalledTimes(1)
})
