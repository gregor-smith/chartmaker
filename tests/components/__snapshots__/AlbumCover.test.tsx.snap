// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`highlighted prop merges styles when false 1`] = `
.emotion-0 {
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  max-width: 100%;
  height: 100%;
  object-fit: contain;
  z-index: 10;
}

.emotion-1 {
  position: relative;
  background: white;
  margin: 0.15rem;
  cursor: -webkit-grab;
  cursor: -moz-grab;
  cursor: grab;
  width: 2rem;
  height: 2rem;
}

.emotion-1:not(:hover) .overlay {
  display: none;
}

<div
  id="test-render-container"
>
  <div
    class="emotion-1"
    draggable="false"
  >
    <div
      class="overlay emotion-0"
    >
      Test children
    </div>
  </div>
</div>
`;

exports[`highlighted prop merges styles when false 2`] = `
.emotion-0 {
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  max-width: 100%;
  height: 100%;
  object-fit: contain;
  z-index: 10;
}

.emotion-1 {
  position: relative;
  background: white;
  margin: 0.15rem;
  cursor: -webkit-grab;
  cursor: -moz-grab;
  cursor: grab;
  width: 2rem;
  height: 2rem;
  opacity: calc(100% / 3 * 2);
}

.emotion-1:not(:hover) .overlay {
  display: none;
}

<div
  id="test-render-container"
>
  <div
    class="emotion-1"
    draggable="false"
  >
    <div
      class="overlay emotion-0"
    >
      Test children
    </div>
  </div>
</div>
`;

exports[`overlayClass prop merges styles 1`] = `
.emotion-1 {
  position: relative;
  background: white;
  margin: 0.15rem;
  cursor: -webkit-grab;
  cursor: -moz-grab;
  cursor: grab;
  width: 2rem;
  height: 2rem;
}

.emotion-1:not(:hover) .overlay {
  display: none;
}

.emotion-0 {
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  max-width: 100%;
  height: 100%;
  object-fit: contain;
  z-index: 10;
  color: red;
}

<div
  id="test-render-container"
>
  <div
    class="emotion-1"
    draggable="false"
  >
    <div
      class="overlay emotion-0"
    >
      Test children
    </div>
  </div>
</div>
`;

exports[`renders with image for named album 1`] = `
.emotion-0 {
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  max-width: 100%;
  height: 100%;
  object-fit: contain;
}

.emotion-1 {
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  max-width: 100%;
  height: 100%;
  object-fit: contain;
  z-index: 10;
}

.emotion-2 {
  position: relative;
  background: white;
  margin: 0.15rem;
  cursor: -webkit-grab;
  cursor: -moz-grab;
  cursor: grab;
  width: 2rem;
  height: 2rem;
}

.emotion-2:not(:hover) .overlay {
  display: none;
}

<div
  id="test-render-container"
>
  <div
    class="emotion-2"
    draggable="true"
    title="Test album"
  >
    <img
      alt="Test album"
      class="emotion-0"
      crossorigin="anonymous"
      src="https://test.com"
    />
    <div
      class="overlay emotion-1"
    >
      Test children
    </div>
  </div>
</div>
`;

exports[`renders without image for placeholder album 1`] = `
.emotion-0 {
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  max-width: 100%;
  height: 100%;
  object-fit: contain;
  z-index: 10;
}

.emotion-1 {
  position: relative;
  background: white;
  margin: 0.15rem;
  cursor: -webkit-grab;
  cursor: -moz-grab;
  cursor: grab;
  width: 2rem;
  height: 2rem;
}

.emotion-1:not(:hover) .overlay {
  display: none;
}

<div
  id="test-render-container"
>
  <div
    class="emotion-1"
    draggable="false"
  >
    <div
      class="overlay emotion-0"
    >
      Test children
    </div>
  </div>
</div>
`;
